enterprise/internal/authz/bitbucketcloud/provider.go

```diff
@@ -0,0 +1,183 @@
+		}
+	}
+
+	return &Provider{
+		urn:      conn.URN,
+		codeHost: extsvc.NewCodeHost(baseURL, extsvc.TypeBitbucketCloud),
+		client:   opts.BitbucketCloudClient,
+		pageSize: 1000,
+		db:       db,
+	}
+}
+
+// ValidateConnection validates that the Provider has access to the Bitbucket Server API
+// with the OAuth credentials it was configured with.
+func (p *Provider) ValidateConnection(ctx context.Context) []string {
+	return nil
+}
+
+func (p *Provider) URN() string {
+	return p.urn
+}
+
+// ServiceID returns the absolute URL that identifies the Bitbucket Server instance
+// this provider is configured with.
+func (p *Provider) ServiceID() string { return p.codeHost.ServiceID }
+
+// ServiceType returns the type of this Provider, namely, "bitbucketCloud".
+func (p *Provider) ServiceType() string { return p.codeHost.ServiceType }
+
+// FetchAccount satisfies the authz.Provider interface.
+func (p *Provider) FetchAccount(ctx context.Context, user *types.User, _ []*extsvc.Account, _ []string) (acct *extsvc.Account, err error) {
+	return nil, nil
+}
+
+// FetchUserPerms returns a list of repository IDs (on code host) that the given account
+// has read access on the code host. The repository ID has the same value as it would be
+// used as api.ExternalRepoSpec.ID. The returned list only includes private repository IDs.
+//
+// This method may return partial but valid results in case of error, and it is up to
+// callers to decide whether to discard.
+//
+// API docs: https://docs.atlassian.com/bitbucket-server/rest/5.16.0/bitbucket-rest.html#idm8296923984
+func (p *Provider) FetchUserPerms(ctx context.Context, account *extsvc.Account, opts authz.FetchPermsOptions) (*authz.ExternalUserPermissions, error) {
+	switch {
+	case account == nil:
+		return nil, errors.New("no account provided")
+	case !extsvc.IsHostOfAccount(p.codeHost, account):
+		return nil, errors.Errorf("not a code host of the account: want %q but have %q",
+			p.codeHost.ServiceID, account.AccountSpec.ServiceID)
+	case account.Data == nil:
+		return nil, errors.New("no account data provided")
+	}
+
+	_, tok, err := bitbucketcloud.GetExternalAccountData(ctx, &account.AccountData)
+	if err != nil {
+		return nil, err
+	}
+	oauthToken := &auth.OAuthBearerToken{
+		Token:        tok.AccessToken,
+		RefreshToken: tok.RefreshToken,
+		Expiry:       tok.Expiry,
+	}
+
+	oauthToken.RefreshFunc = database.GetAccountRefreshAndStoreOAuthTokenFunc(p.db, account.ID, bitbucketcloud.GetOAuthContext(p.codeHost.BaseURL.String()))
+	oauthToken.NeedsRefreshBuffer = 5
+
+	client := p.client.WithAuthenticator(oauthToken)
+	repos, next, err := client.Repos(ctx, nil, "", &bitbucketcloud.ReposOptions{Role: "member"})
+	if err != nil {
+		return nil, err
+	}
+	for next.HasMore() {
+		var nextRepos []*bitbucketcloud.Repo
+		nextRepos, next, err = client.Repos(ctx, next, "", &bitbucketcloud.ReposOptions{Role: "member"})
+		if err != nil {
+			return nil, err
+		}
+		repos = append(repos, nextRepos...)
+	}
```

Comment:
I think we can move the first iteration into the loop as well. This is my idea, not sure how much it adheres to go conventions:
```suggestion
	repos := []*bitbucketcloud.Repo{}
	for next := nil; next == nil || next.HasMore(); {
		var nextRepos []*bitbucketcloud.Repo
		nextRepos, next, err = client.Repos(ctx, next, "", &bitbucketcloud.ReposOptions{Role: "member"})
		if err != nil {
			return nil, err
		}
		repos = append(repos, nextRepos...)
	}
```